### DISCUSSION TITLE
Definition of a 2 materials sample (elastic+viscoelastic)

### DISCUSSION THREAD

**Discussion by AlexSacMorane** [idaholab/moose#28524] [Q&A Modules: Solid mechanics] [answered]
Dear all,

I would like to model with MOOSE a 2 materials sample. This one is divided into one elastic part and one viscoelastic part.

Thanks to the tutorials, I have successfully run simulations with an elastic material only and with a viscoelastic material only. Nevertheless, I have not, combining the two.

The script I am using now is the following or in the attachments [Elastic_ViscoElastic_Parallel_Materials.txt](https://github.com/user-attachments/files/16840897/Elastic_ViscoElastic_Parallel_Materials.txt).

> [Mesh]
>   [generated]
>     type = GeneratedMeshGenerator
>     dim = 2
>     nx = 10
>     ny = 10
>     xmin = -0.5
>     xmax =  0.5
>     ymin = -0.5
>     ymax =  0.5
>   []
> 
>   [elastic]
>     type = SubdomainBoundingBoxGenerator
>     input = generated
>     block_id = 1
>     bottom_left = '-0.5 -0.5 0'
>     top_right = '0 0.5 0'
>   []
>   [viscoelastic]
>     type = SubdomainBoundingBoxGenerator
>     input = elastic
>     block_id = 2
>     bottom_left = '0 -0.5 0'
>     top_right = '0.5 0.5 0'
>   []
> []
> 
> [GlobalParams]
>   displacements = 'disp_x disp_y'
> []
> 
> [Modules/TensorMechanics/Master]
>   add_variables = true
>   generate_output = 'stress_xx stress_xy stress_yx stress_yy strain_xx strain_xy strain_yx strain_yy'
>   [./block1]
>     block = 1
>   [../]
>   [./block2]
>     block = 2
>   [../]
> []
> 
> [BCs]
>   inactive = top_y
>   [bottom_x]
>     type = DirichletBC
>     variable = disp_x
>     boundary = bottom
>     value = 0
>   []
>   [bottom_y]
>     type = DirichletBC
>     variable = disp_y
>     boundary = bottom
>     value = 0
>   []
>   [top_x]
>     type = FunctionDirichletBC
>     variable = disp_x
>     boundary = top
>     function = 1*t
>   []
>   [top_y]
>     type = FunctionDirichletBC
>     variable = disp_y
>     boundary = top
>     function = 0.5*t
>   []
>   [./Periodic]
>     [./per_x_x]
>       variable = disp_x
>       auto_direction = 'x'
>     [../]
>     [./per_y_x]
>       variable = disp_y
>       auto_direction = 'x'
>     [../]
>   [../]
> []
> 
> [Materials]
>   [./elasticity_tensor]
>     type = ComputeIsotropicElasticityTensor
>     youngs_modulus = 1e9
>     poissons_ratio = 0.3
>     bloc = 1
>   [../]
>   [./kelvin_voigt]
>     type = GeneralizedKelvinVoigtModel
>     creep_modulus = '1e9'
>     creep_viscosity = '1'
>     poisson_ratio = 0.3
>     young_modulus = 1e9
>     bloc = 2
>   [../]
>   [./stress_elastic]
>     type = ComputeLinearElasticStress
>     block = 1
>   [../]
>   [./stress_viscoelastic]
>     type = ComputeLinearViscoelasticStress
>     bloc = 2
>   [../]
> []
> 
> [UserObjects]
>   [./update]
>     type = LinearViscoelasticityManager
>     viscoelastic_model = kelvin_voigt
>     bloc = 2
>   [../]
> []
> 
> [Preconditioning]
>   [SMP]
>     type = SMP
>     full = true
>   []
> []
> 
> [Executioner]
>   type = Transient
> 
>   l_max_its  = 100
>   l_tol      = 1e-8
>   nl_max_its = 50
>   nl_rel_tol = 1e-8
>   nl_abs_tol = 1e-8
>   
>   petsc_options_iname = '-pc_type'
>   petsc_options_value = 'lu'
>   
>   end_time = 1
>   dt = 0.01
> []
> 
> [Outputs]
>   exodus = true
>   [console]
>     type = Console
>     execute_on = 'nonlinear'
>   []
> []

The output I get is the following:

> *** ERROR ***
> The following error occurred in the object "MOOSE Problem", of type "FEProblem".
> 
> The following material properties are declared on block 1 by multiple materials:
>   Material Property             Material Objects
>   effective_stiffness           block1_strain elasticity_tensor kelvin_voigt stress_elastic stress_viscoelastic
>   elasticity_tensor             block1_strain elasticity_tensor kelvin_voigt stress_elastic stress_viscoelastic

I am using an old version of Moose, the module is named tensor_mechanics. I am unsure about where the problem comes from. Is there someone able to help me? 
Regards
Alexandre

**Comment by GiudGiud** [comment, marked as answer]
Hello

Materials can be block restricted, which you almost have working.
Instead of `bloc = 1` you want `block = 1` in the `elasticity_tensor` material.
Same for the others.

I think the 'unused parameter warning' would have helped you, but maybe it did not show because of the error

**Reply by AlexSacMorane** [reply]
Thank you, it was indeed an error because of that. I was not aware. Have a great day and thanks for your time.
